src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs

```diff
@@ -20952,32 +20952,406 @@ public class C
+                //         field4.ToString(); // 4
+                Diagnostic(ErrorCode.WRN_NullReferenceReceiver, "field4").WithLocation(22, 9));
+        }
+
+        [Fact, WorkItem("https://github.com/dotnet/roslyn/issues/56256")]
+        public void MemberNotNull_LocalFunction_WithinAnonymousFunction()
+        {
+            var c = CreateNullableCompilation(["""
+using System.Diagnostics.CodeAnalysis;
+public class C
+{
+    public string field1;
+    public string? field2;
+    public string field3;
+    public string? field4;
+
+    public C()
+    {
+        System.Action a = delegate()
+        {
+            init();
+
+            [MemberNotNull(nameof(field1), nameof(field2))]
+            void init() => throw null!;
+        };
+
+        field1.ToString(); // 1
+        field2.ToString(); // 2
+        field3.ToString(); // 3
+        field4.ToString(); // 4
     }
 }
-", MemberNotNullAttributeDefinition }, parseOptions: TestOptions.Regular9);
+""", MemberNotNullAttributeDefinition]);
 
-            // Note: the local function is not invoked on this or base
             c.VerifyDiagnostics(
-                // (13,9): warning CS8602: Dereference of a possibly null reference.
+                // (19,9): warning CS8602: Dereference of a possibly null reference.
                 //         field1.ToString(); // 1
-                Diagnostic(ErrorCode.WRN_NullReferenceReceiver, "field1").WithLocation(13, 9),
-                // (14,9): warning CS8602: Dereference of a possibly null reference.
+                Diagnostic(ErrorCode.WRN_NullReferenceReceiver, "field1").WithLocation(19, 9),
+                // (20,9): warning CS8602: Dereference of a possibly null reference.
                 //         field2.ToString(); // 2
-                Diagnostic(ErrorCode.WRN_NullReferenceReceiver, "field2").WithLocation(14, 9),
-                // (15,9): warning CS8602: Dereference of a possibly null reference.
+                Diagnostic(ErrorCode.WRN_NullReferenceReceiver, "field2").WithLocation(20, 9),
+                // (21,9): warning CS8602: Dereference of a possibly null reference.
                 //         field3.ToString(); // 3
-                Diagnostic(ErrorCode.WRN_NullReferenceReceiver, "field3").WithLocation(15, 9),
-                // (16,9): warning CS8602: Dereference of a possibly null reference.
+                Diagnostic(ErrorCode.WRN_NullReferenceReceiver, "field3").WithLocation(21, 9),
+                // (22,9): warning CS8602: Dereference of a possibly null reference.
                 //         field4.ToString(); // 4
-                Diagnostic(ErrorCode.WRN_NullReferenceReceiver, "field4").WithLocation(16, 9)
-                );
+                Diagnostic(ErrorCode.WRN_NullReferenceReceiver, "field4").WithLocation(22, 9));
+        }
+
+        [Fact, WorkItem("https://github.com/dotnet/roslyn/issues/56256")]
+        public void MemberNotNull_LocalFunction_WithinExtensionMethod()
```

Comment:
It looks like an extension method is not used in this test.

Reply:
I couldn't come up with any interesting test involving extension methods after all. Removed